### DISCUSSION TITLE
Lesson 9: testPerformUpkeepCanOnlyRunIfCheckUpkeepIsTrue

### DISCUSSION THREAD

**Discussion by sharonphiliplima** [Cyfrin/foundry-full-course-cu#468] [Q&A]
```
`Hey!
function testPerformUpkeepCanOnlyRunIfCheckUpkeepIsTrue() public {
        //Arrange
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee};
        (bool upkeepNeeded, ) = raffle.checkUpkeep("");
        assert(!upkeepNeeded);
        //Act / Assert
        raffle.performUpkeep("");
        vm.expectRevert(Raffle.Raffle__UpkeepNotNeeded.selector);
    }
```
    
    The above test failed!
    
    Code: 
    
    function performUpkeep(bytes calldata /* performData */) external {
        (bool upkeepNeeded, ) = checkUpkeep("");
        if (!upkeepNeeded) {
            revert Raffle__UpkeepNotNeeded(
                address(this).balance,
                s_players.length,
                uint256(s_raffleState)
            );
        }`

Please explain how. 

This test runs successfully for the code: 
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/69852010/0b26e632-da34-400c-924d-576e22e9551c)

**Comment by Naman1729** [comment]
Hey @sharonphiliplima , 
Answer 1.
You can do this on your test :
```solidity
    function testPerformUpkeepCanOnlyRunIfCheckUpkeepIsTrue() public {
        // Arrange
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
        (bool upkeepNeeded, ) = raffle.checkUpkeep("");
        assert(!upkeepNeeded);

        // Act / Assert
        vm.expectRevert(abi.encodeWithSelector(Raffle.Raffle__UpkeepNotNeeded.selector, 10000000000000000, 1, 0));
        raffle.performUpkeep("");
    }
```
Here, we use expectRevert before that function call which reverting something. so, Raffle__UpkeepNotNeeded have three arguments "**currentBalance, numPlayers, raffleState**"  We put these arguments and encode our revert statement which effectively handles our check. Syntax of the expectRevert statement which contain arguments :
```solidity
vm.expectRevert(abi.encodeWithSelector(Raffle.Raffle__UpkeepNotNeeded.selector, 10000000000000000, 1, 0));
```

Question 2.
vm.wrap(TIME) is use to change the timestamp equal to the argument. i.e after vm.wrap(TIME) the timestamp become equal to the TIME( block.timestamp=TIME). Here, we use vm.wrap(block.timestamp + interval + 1) .because, our Contract decide the lottery winner after the interval from current block.timestamp. so, we have to skip our interval time and go to that time when winner selection is completed (Just Like Time Travel 🕓🚀🕢).
similarly, vm.roll(block.number + 1) is used to change the block number. here, we add only 1 to go to the next block number.

**Comment by theirrationalone** [comment]
Hey take a look at this, [i explained](https://github.com/Cyfrin/foundry-full-course-f23/discussions/535#discussioncomment-6636999)....

**Comment by theirrationalone** [comment]
> Hey take a look at this, [i explained](https://github.com/Cyfrin/foundry-full-course-f23/discussions/535#discussioncomment-6636999)....
vm.roll -> It is used to mine the current block, This means we want to be clear & sure that nothing left to do... No any transaction is pending, It's all clear and transaction is added to The Blockchain.

Remember One block can have several transactions, Yours is one among those in that block. We need a miner or validator(Now it's the Truth) to validate & check our transaction. Here foundry itself  the validator/miner who mines/rolls/validates our transaction by using "vm.roll" facility.

```solidity
vm.roll(block.number + 1);
```

block.number: is the current block to mine/validate.
+ 1 : Added to make sure nothing remained pending, So we can perform next transactions easily.
